### PR TITLE
Add Flow control function IF(expr1, expr2, expr3)

### PR DESCRIPTION
**Pull request by harold-wang**
*Issue #861 , if available:*

*Description of changes:*

Add flow control function IF(expr1, expr2, expr3)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 1]
```diff
@@ -149,4 +163,8 @@ public static ExprValue exprNullIf(ExprValue v1, ExprValue v2) {
     return v1.equals(v2) ? LITERAL_NULL : v1;
   }
 
+  public static ExprValue exprIf(ExprValue v1, ExprValue v2, ExprValue v3) {
+    return v1.value().equals(LITERAL_TRUE.value()) ? v2 : v3;
```
Comment (dai-chen) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:167]: For `v1`, I think we need to do null/missing check? Please check MySQL behavior and if ours is the same.

And `ExprBooleanValue` already implements `equals()` so just do `LITERAL_TRUE.equals(v1)`?

File: docs/user/dql/functions.rst [hunk 2]
```diff
@@ -1982,6 +1982,40 @@ Example::
...
+------
+
+Description
+>>>>>>>>>>>
+
+Specifications:
+
+1. IF(ES_TYPE0, ES_TYPE1, ES_TYPE2) -> ES_TYPE1 or ES_TYPE2
+
+Usage: return first parameter is true, return second parameter, otherwise return third one.
```
Comment (dai-chen) [inline docs/user/dql/functions.rst:1995]: np: "`if` first parameter is true ..."

**Comment by dai-chen** [comment]
A question: should the function signature register with `UNKNOWN` be removed too? https://github.com/opendistro-for-elasticsearch/sql/pull/990/files#diff-ee3435d38230141befb55d9dce13674513b55496848176905e88e44ecaeffea5R126-R129

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 1 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:167]: Thanks @dai-chen, you are correct, we need to do null/missing check. let me update both.

https://www.w3resource.com/mysql/control-flow-functions/if-function.php
IF() function
MySQL IF() takes three expressions and if the first expression is true, not zero and not NULL, it returns the second expression. Otherwise, it returns the third expression.

Depending on the context in which it is used, it returns either numeric or string value.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 3]
```diff
@@ -100,6 +99,25 @@ private static FunctionResolver isNotNull() {
...
+
+    List<SerializableFunction<FunctionName, org.apache.commons.lang3.tuple.Pair<FunctionSignature,
+            FunctionBuilder>>> functionsOne = typeList.stream().map(v ->
+            impl((UnaryPredicateOperator::exprIf), v, BOOLEAN, v, v))
+            .collect(Collectors.toList());
+
+    List<SerializableFunction<FunctionName, org.apache.commons.lang3.tuple.Pair<FunctionSignature,
+            FunctionBuilder>>> functionsTwo = typeList.stream().map(v ->
+            impl((UnaryPredicateOperator::exprIf), v, UNKNOWN, v, v))
+            .collect(Collectors.toList());
```
Comment (dai-chen) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:114]: Same as my previous question for line 130-133, why is this required?

File: docs/user/dql/functions.rst [hunk 4]
```diff
@@ -1982,6 +1982,40 @@ Example::
     | True          | False         |
     +---------------+---------------+
 
+IF
+------
```
Comment (dai-chen) [inline docs/user/dql/functions.rst:1986]: could you update PPL function doc too?

File: docs/user/dql/functions.rst [hunk 5]
```diff
@@ -1982,6 +1982,40 @@ Example::
...
+
+Description
+>>>>>>>>>>>
+
+Specifications:
+
+1. IF(ES_TYPE0, ES_TYPE1, ES_TYPE2) -> ES_TYPE1 or ES_TYPE2
```
Comment (penghuo) [inline docs/user/dql/functions.rst:1993]: Is there any reason we use ES_TYPE0/1/2? In my opinion, IF(expr, expr1, expr2) is more clear definition?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 3 above)
Comment (penghuo) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:114]: could we consider the solution which could for case...when and if. i think they both control flow statement which may not fit into our function defintion.

Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java]: Good idea.

**Review by dai-chen — APPROVED** [review]

**Review by penghuo — APPROVED** [review]

**Comment by harold-wang** [comment]
Todo issue: merge if function with case when function.
